### PR TITLE
Add `std.math.isPowerOf2()`. Supports floating-point and integers.

### PR DESCRIPTION
**Pull request by tsbockman**
This is intended to supersede [PR #4308](https://github.com/dlang/phobos/pull/4308), which adds `std.math.isPowerOf2()`. Improvements:
* Support floating-point numbers.
* Corrected handling of signed integers.
* Reduced template bloat, by marking `x` as `const`.

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -7607,3 +7607,63 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
         assert(truncPow2(T.init).isNaN);
     }
 }
+
+
+/**
+Quickly check whether a number is a power of two.
```
Comment (JackStouffer) [inline std/math.d]: > Quickly

Redundant. You would expect the standard library's implementation to be quick.

File: std/math.d [hunk 2]
```diff
@@ -7607,3 +7607,132 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
...
+    {
+        static if (isSigned!X)
+            alias W = typeof(x + 0);
+        else
+            alias W = typeof(x + 0u); // For ubyte and ushort W should be uint
+
+        static if (is(X == W))
+        {
+            static if (isSigned!X)
+                return ((x & -x) == x) && (x > -x);
```
Comment (tsbockman) [inline std/math.d]: The signed algorithm uses two extra instructions in order to correctly handle `int.min` (as well as `0`). It is still branch-free, at least on `x86_64` with GDC.

`std.allocator` and the like use an unsigned `X`, and so will continue to use Matthias Bentrup's slightly faster algorithm, below.

**Comment by tsbockman** [comment]
@JackStouffer I have now added a commit to update `std.allocator` (and a few other places). Please review.

File: std/math.d [hunk 3]
```diff
@@ -7607,3 +7607,132 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
...
+            else
+                return (x & -x) > (x - 1); // Credit: Matthias Bentrup
```
Comment (ibuclaw) [inline std/math.d]: Who is Matthias Bentrup and why is he credited for a commo knowledge algorithms? :-)

File: std/math.d [hunk 4]
```diff
@@ -7607,3 +7607,132 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
...
+bool isPowerOf2(X)(const X x) pure @safe nothrow @nogc
+    if (isNumeric!X)
+{
+    static if (isFloatingPoint!X)
+    {
+        if (__ctfe)
+        {
+            // TODO: Simplify this as more floating-point functions are made to work at CTFE
+            const isFinite = -real.infinity < x && x < real.infinity;
+            if (!isFinite)
```
Comment (ibuclaw) [inline std/math.d]: Use isFinite function?

**Comment by tsbockman** [comment]
@ibuclaw 
> [Wouldn't this fall under the too trivial for phobos category?](https://github.com/dlang/phobos/pull/4308#issuecomment-219268922)

* Phobos already contains two or three independent implementations of this function for integers; this PR just consolidates them and makes `isPowerOf2()` public.

* When people try to implement this themselves, they often mishandle `0`, and pretty much always mishandle `int.min`.

* This PR also supports floating-point, which is harder to get right.

* Making the floating-point implementation work at CTFE sadly is not trivial at the moment, because so much of `std.math` is off-limits.

File: std/math.d (hunk 4 above)
Comment (tsbockman) [inline std/math.d]: Pull [#4328](https://github.com/dlang/phobos/pull/4328), and I will.

File: std/math.d (hunk 3 above)
Comment (tsbockman) [inline std/math.d]: This PR is a continuation of [#4308](https://github.com/dlang/phobos/pull/4308), which was based on the private `isPowerOf2()` in `std.allocator`.

**andralex** credited him for suggesting that algorithm in https://github.com/dlang/phobos/commit/5adf023876270287811dc0801b0ffd9811c05e55 following [a forum discussion](http://forum.dlang.org/post/mjiald$2l73$1@digitalmars.com).

**Comment by 9il** [comment]
Needs tests for FP numbers around `-F.max` and `F.max`. Both CT and RT. Also needs a proof that CT implementation is correct

**Comment by 9il** [comment]
> This PR also supports floating-point, which is harder to get right.

This is not so hard.

> Making the floating-point implementation work at CTFE sadly is not trivial at the moment, because so much of std.math is off-limits.

We don't need CTFE-able math until CTFE-able unions.

**Comment by tsbockman** [comment]
@9il 
> Needs tests for FP numbers around `-F.max` and `F.max`. 

Done.

> Both CT and RT. Also needs a proof that CT implementation is correct

The CTFE path is covered by the unit test now, also.

> This is not so hard.

Most people are really bad about covering all the corner cases with floating-point (and integer, for that matter). The purpose of the standard library is not only to save people effort, but also to promote the usage of *correct*, properly tested algorithms.

> We don't need CTFE-able math until CTFE-able unions.

I have a specific use case in mind for CTFE `isPowerOf2!real()`. However, since you pointed out in another thread that `if(__ctfe)` blocks inlining - perhaps I should just remove the CTFE support from this for now?

(I can sort of do what I want just with `isPowerOf2!long()`, so this part isn't strictly necessary.)

**Comment by 9il** [comment]
> I have a specific use case in mind for CTFE isPowerOf2!real(). However, since you pointed out in another thread that if(__ctfe) blocks inlining - perhaps I should just remove the CTFE support from this for now?

Yes

**Comment by tsbockman** [comment]
@9il I have removed the `if (__ctfe)` block.

**Comment by tsbockman** [comment]
Rebased.

**Comment by tsbockman** [comment]
Rebased, again.

**Comment by tsbockman** [comment]
Ping @andralex .

**Comment by JackStouffer** [comment]
If you want @andralex's opinion you're going to have to email him

**Comment by tsbockman** [comment]
> you're going to have to email him

I did so earlier today.

File: std/math.d [hunk 5]
```diff
@@ -7619,3 +7619,116 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
...
+*/
+bool isPowerOf2(X)(const X x) pure @safe nothrow @nogc
+    if (isNumeric!X)
+{
+    static if (isFloatingPoint!X)
+    {
+        int exp;
+        const X sig = frexp(x, exp);
+
+        return (exp != int.min) && (sig is cast(X)0.5L);
```
Comment (andralex) [inline std/math.d]: space after `cast(X)` (won't enforce this now)

**Comment by andralex** [comment]
Good work, thx.

**Comment by andralex** [comment]
Auto-merge toggled on

File: std/math.d (hunk 5 above)
Comment (tsbockman) [inline std/math.d]: > (won't enforce this now)

Thank you. Let's please just get anything like this that *will* be enforced into the Travis CI build; it makes the contribution process much less frustrating.

Comment (yebblies) [inline std/math.d]: Other way around actually, no space after cast.
